src/svc_xprt.c

```diff
@@ -338,16 +338,11 @@ svc_xprt_shutdown()
 		t = &svc_xprt_fd.xt.tree[p_ix];
 
 		rwlock_wrlock(&t->lock);	/* t WLOCKED */
-		n = opr_rbtree_first(&t->t);
-		while (n != NULL) {
+		while ((n = opr_rbtree_first(&t->t))) {
 			rec = opr_containerof(n, struct rpc_dplx_rec, fd_node);
-			n = opr_rbtree_next(n);
 
 			/* prevent repeats, see svc_xprt_clear() */
-			rpc_dplx_rli(rec);
```

Comment:
I don't think this is safe.  There appears to be other removals that are only protected by the dplx lock, not the partition lock.  I think what you should do (and what I understood we had discussed yesterday) is to drop the partition lock before taking the dplx lock, then take it again after releasing the dplx lock.  Unfortunately, looking at this, that may not work either, since there is no refcount on the rec, so it can be freed while neither lock is held.  I think we're going to have to rethink this.

Reply:
No. There are only 2 removals, and they are both in this file. The comment points you at the other in svc_xprt_clear(). I actually wrote the two additional lines for dropping the lock around SVC_DESTROY(), and then removed them again.

If another task calls SVC_DESTROY() too and passes the opr_rbtree_node_valid() via svc_xprt_clear(), we don't want it to proceed with the attempted removal. It has to stall by lock until this is all over. I'll add a comment explaining.

As we discussed yesterday, the fd_node in the transport should be controlled by the partition lock, not the duplex lock. The duplex lock (and _FLAG_RELEASING) is used in the SVC_DESTROY() path to ensure only one svc_xprt_clear() can be called per xprt.

Also I carefully looked at the rbtree code, and there was no need for atomic there, as that gen = 0 code is always under partition lock.

Already refcounted fd_node. There is only one per xprt. The xprt begins with xp_refs = 1 for this xprt list entry, that's the extra one decremented by SVC_DESTROY(). I'll add a comment explaining.

Reply:
Okay, I misread removals from other trees, protected by that lock.  This one should be safe.

Reply:
I put in the two additional lines for dropping the lock around SVC_DESTROY() again, because we really don't care about performance during shutdown. Happy?